File: ALGORITHM/Greedy/MST-Kruskal.py
class DisjointSet:
    def __init__(self):
        self.parent = {}
        self.rank = {}

    def make_set(self, vertex):
        """
        创建一个新集合，只包含vertex。
        """
        self.parent[vertex] = vertex
        self.rank[vertex] = 0

    def find_set(self, vertex):
        """
        查找vertex所属的集合的代表元素（根节点）。
        使用路径压缩优化。
        """
        if self.parent[vertex] != vertex:
            self.parent[vertex] = self.find_set(self.parent[vertex])
        return self.parent[vertex]

    def union(self, u, v):
        """
        合并包含u和v的两个集合。
        使用按秩合并优化。
        """
        root_u = self.find_set(u)
        root_v = self.find_set(v)

        if root_u != root_v:
            if self.rank[root_u] > self.rank[root_v]:
                self.parent[root_v] = root_u
            elif self.rank[root_u] < self.rank[root_v]:
                self.parent[root_u] = root_v
            else:
                self.parent[root_v] = root_u
                self.rank[root_u] += 1

File: ALGORITHM/Greedy/MST-Kruskal.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

// 定义边的结构体
struct Edge
{
    int u, v, weight;
};

// 定义并查集
class DisjointSet
{
private:
    vector<int> parent, rank;

public:
    DisjointSet(int n)
    {
        parent.resize(n);
        rank.resize(n, 0);
        // 初始化每个元素的父节点为自身
        for (int i = 0; i < n; ++i)
            parent[i] = i;
    }

    // 查找元素所属的集合（根节点）
    int find(int u)
    {
        if (parent[u] != u)
            parent[u] = find(parent[u]);
        return parent[u];
    }

    // 合并两个集合
    void merge(int u, int v)
    {
        int pu = find(u), pv = find(v);
        if (pu != pv)
        {
            if (rank[pu] < rank[pv])
                swap(pu, pv); // 如果pv更深，那么将pv和pu的值互换
            parent[pv] = pu;  // pu是最深的那个
            if (rank[pu] == rank[pv])
                rank[pu]++; // 如果相等，那么pu的深度增加
        }
    }
};

// Kruskal算法
vector<Edge> kruskal(vector<Edge> &edges, int V)
{
    vector<Edge> MST;
    // 按边的权重升序排序
    sort(edges.begin(), edges.end(), [](const Edge &a, const Edge &b)
         { return a.weight < b.weight; });

    DisjointSet ds(V);

    for (const auto &edge : edges)
    {
        int u = edge.u, v = edge.v, weight = edge.weight;
        if (ds.find(u) != ds.find(v))
        {
            MST.push_back({u, v, weight});
            ds.merge(u, v);
        }
    }

    return MST;
}

int main()
{
    int V, E;
    cout << "Enter number of vertices: ";
    cin >> V;
    cout << "Enter number of edges: ";
    cin >> E;

    vector<Edge> edges(E);
    cout << "Enter edges in the format (u v weight):" << endl;
    for (int i = 0; i < E; ++i)
    {
        cin >> edges[i].u >> edges[i].v >> edges[i].weight;
    }

    vector<Edge> MST = kruskal(edges, V);

    cout << "Minimum Spanning Tree:" << endl;
    for (const auto &edge : MST)
    {
        cout << edge.u << " - " << edge.v << " : " << edge.weight << endl;
    }

    return 0;
}
